### ISSUE TITLE
Add Cloud Run (k8s serverless) support

### ISSUE THREAD

**Issue by dunglas** [api-platform/api-platform#1107]
An API Platform project should be easily deployable in a Cloud Run environment: https://cloud.google.com/run/

We should provide the ability to do it out of the box, as done for the Helm chart.

The main issue is that we need to disable Varnish support when using Cloud Run.

By the way, it should be nice to deploy the demo of API Platform using Cloud Run, to reduce our hosting costs. We should use a managed Varnish (such as Fastly), or add support for CloudFlare Workers for the cache.

**Comment by nietzscheson** [comment]
Good idea!, for php there is a library that implemente a serveless services on AWS lambda: 
- https://bref.sh/ 
- https://bref.sh/docs/frameworks/symfony.html

**Comment by vbourdeix** [comment]
I totally agree as right now, the default mode pushed by API Platform is to run it on a Kubernetes cluster which can get quite expensive for a test project.

Is there a documentation somewhere giving an example configuration to replace the default Varnish one for a Fastly cache ? It would be a first step to help people deploy on cloud run.

**Comment by akira28** [comment]
We are planning to move one of our API to CloudRun in the near future. We also plan to use Fastly, and as far as I understand Fastly is not supported out of the box (the cache invalidator use BAN that's not supported by Fastly #727 ).
Should I wait for an official implementation or it is not a priority at the moment?

**Comment by comxd** [comment]
About Mercure, it need a single node (in free/community version) but currently the APIP Docker base image include Caddy + Mercure. Cloud Run can scale and create more than one instance, in this case some Mercure clients no longer receive the messages causing an inconsistency between clients/users.
That's right?

**Comment by dunglas** [comment]
Yes, the free version of Mercure only support using a single node (to use a cluster, you need the paid version or to write your own transport).

**Comment by comxd** [comment]
Thank's @dunglas for the confirmation.

> According to a benchmark made by Glory4Gamers, the open-source version of the Mercure.rocks hub is able to 40k concurrent connections on a single EC2 t3.micro instance. It's even possible to handle way more connections by using the HA version of the Mercure.rocks Hub.
Source: https://mercure.rocks/docs/hub/load-test

IMHO 40k is enough for a lot of apps (!) businesses users that make money (and want premium support) using Mercure are welcomed to support the development, I'm totally agreed with that ✔️

APIP does not require Varnish and/or Mercure to be functional, people can use a managed services if needed. The only important thing to remember about the free version of Mercure is to use it on a single node/container.
What about Vulcain?

About GCR, GCR instances are limited to 512MB by default (configurable). Depending on this configured memory limit, we need to ensure that PHP is not using full instance memory, but is depending on the application/code + GCR scaling configuration. @see https://cloud.google.com/run/docs/configuring/memory-limits?hl=en

From the doc (https://cloud.google.com/run/docs/configuring/memory-limits?hl=en#optimizing):
> "You can determine the peak memory requirement for a service using the following: (Standing Memory) + (Memory per Request) * (Service Concurrency)"

Regarding this issue, and globally concerning Mercure in production, tasks to resolve this issue seems to be:

1. Create a dedicated Docker image based on current (or a new stage?)
2. Remove Varnish + Mercure hub from the Docker base image
3. I have encountered some issues with `setfacl` using GCR, remove related `facl` commands in `docker-entrypoint.sh` working well, doesn't that make any security issues?
4. Update the APIP doc

**Comment by soyuka** [comment]
Closing per #2987 — this repository is becoming a CLI installer. The previous Docker/template distribution is being superseded by the installer + upstream [symfony-docker](https://github.com/dunglas/symfony-docker). If the issue still reproduces with the new setup, please open it on `dunglas/symfony-docker` (for Docker/Caddy/FrankenPHP issues) or `api-platform/core` (for framework issues).
